### PR TITLE
Modify run_tests to pass for unknown compilers

### PR DESCRIPTION
**Pull request by thunderboltsid**
Default compiler to Python2.7 in case of unknown compiler
if the travis flag is present.

This behaviour is required as a workaround to #7934

### CODE REVIEW

**Comment by thunderboltsid** [comment]
@nathanielmanistaatgoogle please take a look and add necessary tags.

File: tools/run_tests/run_tests.py [hunk 1]
```diff
@@ -884,6 +884,38 @@ def runs_per_test_type(arg_str):
         msg = '\'{}\' is not a positive integer or \'inf\''.format(arg_str)
         raise argparse.ArgumentTypeError(msg)
 
+_COMPILER_CHOICES = ['default',
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: Never write a list literal when a tuple literal will do, especially for a constant.

File: tools/run_tests/run_tests.py [hunk 2]
```diff
@@ -884,6 +884,38 @@ def runs_per_test_type(arg_str):
...
+                     'gcc4.4', 'gcc4.6', 'gcc4.9', 'gcc5.3',
+                     'clang3.4', 'clang3.5', 'clang3.6', 'clang3.7',
+                     'vs2010', 'vs2013', 'vs2015',
+                     'python2.7', 'python3.4',
+                     'node0.12', 'node4', 'node5',
+                     'coreclr'],
+
+
+def is_travis_enabled():
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: Add a leading underscore. Privatize by default, only remove leading underscores when code elements need to be available outside the module in which they are defined.

File: tools/run_tests/run_tests.py [hunk 3]
```diff
@@ -884,6 +884,38 @@ def runs_per_test_type(arg_str):
...
+  for arg in sys.argv:
+    if arg == "--travis" or arg == "-t":
+      return True
+  return False
+
+
+def compiler_choices(compiler):
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: Leading underscore.

Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: Add an octothorpe-delimited implementation comment explaining what this is, why it exists, and `TODO(https://github.com/grpc/grpc/issues/<number>):` hyperlinking to the issue filed that, when resolved, would obviate the need for it.

File: tools/run_tests/run_tests.py [hunk 4]
```diff
@@ -884,6 +884,38 @@ def runs_per_test_type(arg_str):
...
+    if compiler in _COMPILER_CHOICES:
+      return compiler
+    else:
+      return 'python2.7'
+  else:
+    if compiler in _COMPILER_CHOICES:
+      return compiler
+    else:
+      raise argparse.ArgumentTypeError(str(compiler) + " is not a valid "
+                                                       "compiler option. "
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: Avoid implicit string concatenation whenever possible. How about
```
raise argparse.ArgumentTypeError(
    '{} is not a valid compiler option; please use one from {}'.format(
        compiler, _COMPILER_CHOICES))
```
?

**Comment by thunderboltsid** [comment]
@nathanielmanistaatgoogle Ready for another round of reviews.

File: tools/run_tests/run_tests.py [hunk 5]
```diff
@@ -884,6 +884,41 @@ def runs_per_test_type(arg_str):
...
+
+
+def _is_travis_enabled():
+  for arg in sys.argv:
+    if arg == "--travis" or arg == "-t":
+      return True
+  return False
+
+
+# This method is required as a workaround for issue #7934. This would not
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: Minor typographic nitpick, but:
```
# TODO(https://github.com/grpc/grpc/issues/7934): <multi-line
# description of
# work TO be DOne>
```
.

File: tools/run_tests/run_tests.py [hunk 6]
```diff
@@ -884,6 +884,41 @@ def runs_per_test_type(arg_str):
...
+  if _is_travis_enabled():
+    if compiler in _COMPILER_CHOICES:
+      return compiler
+    else:
+      return 'python2.7'
+  else:
+    if compiler in _COMPILER_CHOICES:
+      return compiler
+    else:
+      raise argparse.ArgumentTypeError("{} is not a valid compiler option; "
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: Really really avoid implicit string concatenation whenever you possibly can. Did line-wrapping at the opening parenthesis not do the trick here? What if you drop the "please"?

**Comment by thunderboltsid** [comment]
@nathanielmanistaatgoogle made changes accordingly. Please take a look again.

File: tools/run_tests/run_tests.py [hunk 7]
```diff
@@ -884,6 +884,40 @@ def runs_per_test_type(arg_str):
...
+
+
+def _is_travis_enabled():
+  for arg in sys.argv:
+    if arg == "--travis" or arg == "-t":
+      return True
+  return False
+
+
+# TODO (https://github./grpc/grpc/issues/7934): This workaround would not
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: My apologies for not noticing last review round: this text should describe an action to be done, in this case "remove this workaround when the testing infrastructure supports...".

Comment (thunderboltsid) [inline tools/run_tests/run_tests.py]: @nathanielmanistaatgoogle fixed.

Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: Fixed and pushed? Or just fixed?

**Comment by nathanielmanistaatgoogle** [comment]
@nicolasnoble: please take a look at the approach here and share your thoughts?

File: tools/run_tests/run_tests.py (hunk 7 above)
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: Whoops; my tired eyes and the dangers of switching back and forth between two code review systems. Nevermind.

File: tools/run_tests/run_tests.py [hunk 8]
```diff
@@ -884,6 +884,42 @@ def runs_per_test_type(arg_str):
...
+# have been necessary if the testing infrastructure provided provisions for
+# supporting different configurations for different branches on the repository.
+# Remove this workaround once test infrastructure supports configurations for
+# different branches.
+def _compiler_choices(compiler):
+  if _is_travis_enabled():
+    if compiler in _COMPILER_CHOICES:
+      return compiler
+    else:
+      return 'python2.7'
```
Comment (nicolasnoble) [inline tools/run_tests/run_tests.py]: So, this is a bit more complex than that, because this should be used by more than just Python.

We rather need to return "none" or "freepass" or "void" or whatever wording you think is appropriate. Then later, in the _build_and_run function, we need to check for this, and build nothing and test nothing, but return a valid status code, and an empty junit test file (the xml_report basically).

**Comment by thunderboltsid** [comment]
@nathanielmanistaatgoogle Please take a look.

File: tools/run_tests/run_tests.py [hunk 9]
```diff
@@ -155,6 +155,8 @@ def configure(self, config, args):
                                                                        self.args.compiler)
 
   def test_specs(self):
+    if self.args.compiler == 'freepass' and self.args.travis:
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: Rather than the string literal `'freepass'` appearing several times throughout this module, make it a module-scope constant (`_FREE_PASS`). If possible, make its value just a plain `object` (`_FREE_PASS = object()`) and use `is` rather than `==` when testing for it.

This will help the code take on a shape that better communicates that the free pass is not just another compiler choice.

File: tools/run_tests/run_tests.py [hunk 10]
```diff
@@ -155,6 +155,8 @@ def configure(self, config, args):
...
+      return []
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: In nearly every other code review these days I say "never write a list where a tuple will do".

This file might just have to wait, it might look weird to be inconsistent and it's much too much of a change to make in the same pull request as something semantically unrelated.

**Comment by jtattermusch** [comment]
I agree with the problem statement in #7934 but I don't think this approach is the right solution. It is confusing and only makes things more error prone. FTR, I am against merging this change.

**Comment by nathanielmanistaatgoogle** [comment]
@jtattermusch: welcome back and glad to have you involved. If this change isn't the way forward, **tell us more about what is**?

**Comment by jtattermusch** [comment]
I *think* the way forward is to have the list of configurations to run in a matrix checked into the repository - that way, if some branch doesn't contain the support of given compiler, it also won't attempt to run it because the test config and actual code will stay in sync.
We currently don't have a very good way to achieve what I described (because of Jenkins wierdness where list of configurations lives in Jenkins job config and not in the repo), but over time we will need to fix this somehow (in a less ad-hoc way than what's attempted in this PR).

One possible remedy is the work done in  https://github.com/grpc/grpc/pull/8141 (still in review and not clear if we are going to use it).

**Comment by nathanielmanistaatgoogle** [comment]
@jtattermusch: how certain are you that working toward a perfect future should block an improvement doable today? [Pull request 8141](https://github.com/grpc/grpc/pull/8141) doesn't conflict with this pull request; why shouldn't we do both?

We've been working on this change for nearly a month and we think it will let us have Python 3.5 and Python 3.6 tests running within days of submission - would taking another path bring us to that objective that quickly?

Would adopting this change interfere with later work being done the "right" way, however the "right" way is decided?

I don't think we'd much mind if this change were merged today but went on to be entirely obviated a few months from now.

**Comment by jtattermusch** [comment]
Ok, so to summarize, your problem is that master branch has support for python 3.5 and 3.6 (not available in v1.0.x) and our Jenkins PR job tries to run python 3.5 and 3.6 when you create a PR against 1.0.x as well (and you need that status to be green because it's required for merging the PR). The same problem doesn't exist for the continuous run for master and 1.0.x branch (non-PR) because these are separate Jenkins jobs and can be set accordingly. Is that right?

I don't want to slow down your progress but as of now, I don't think I am happy with the way the change is made (the diff seems confusing for what you are trying to achieve). I'll take a closer look and try to advise some improvements.

File: tools/run_tests/run_tests.py [hunk 11]
```diff
@@ -155,6 +156,8 @@ def configure(self, config, args):
                                                                        self.args.compiler)
 
   def test_specs(self):
+    if self.args.compiler is _FREE_PASS and self.args.travis:
```
Comment (thunderboltsid) [inline tools/run_tests/run_tests.py]: We are currently seeing this with just Python3.5, 3.6, and PyPy. But I don't see how other languages won't be affected by the same problem in future. The original fix was Python focussed but the conclusion was that the change is better made to all languages for consistent behaviour.

File: tools/run_tests/run_tests.py [hunk 12]
```diff
@@ -77,6 +76,8 @@ def platform_string():
 
 _DEFAULT_TIMEOUT_SECONDS = 5 * 60
 
+_FREE_PASS = object()
```
Comment (thunderboltsid) [inline tools/run_tests/run_tests.py]: Added one. Please take a look.

File: tools/run_tests/run_tests.py [hunk 13]
```diff
@@ -65,7 +65,6 @@
   'GRPC_VERBOSITY': 'DEBUG',
 }
 
-
```
Comment (jtattermusch) [inline tools/run_tests/run_tests.py]: nit: revert remove line.

File: tools/run_tests/run_tests.py [hunk 14]
```diff
@@ -878,6 +957,43 @@ def runs_per_test_type(arg_str):
...
+      return True
+  return False
+
+
+# TODO (https://github./grpc/grpc/issues/7934): This workaround would not
+# have been necessary if the testing infrastructure provided provisions for
+# supporting different configurations for different branches on the repository.
+# Remove this workaround once test infrastructure supports configurations for
+# different branches.
+def _compiler_choices(compiler):
```
Comment (jtattermusch) [inline tools/run_tests/run_tests.py]: Add func comment """Validates value of --compiler argument passed on commandline."""  or sth in that sense.

File: tools/run_tests/run_tests.py [hunk 15]
```diff
@@ -878,6 +957,43 @@ def runs_per_test_type(arg_str):
...
+                     'gcc4.4', 'gcc4.6', 'gcc4.9', 'gcc5.3',
+                     'clang3.4', 'clang3.5', 'clang3.6', 'clang3.7',
+                     'vs2010', 'vs2013', 'vs2015',
+                     'python2.7', 'python3.4',
+                     'node0.12', 'node4', 'node5',
+                     'coreclr',
+                     _FREE_PASS,)
+
+
+def _is_travis_enabled():
```
Comment (jtattermusch) [inline tools/run_tests/run_tests.py]: ok, just noticed this function. really dirty. Arguments are already being parsed by argparse and we should really stick with that.

What you could do:
just accept anything as --compiler value (and perhaps add list of allowed values to the help string), and right after parsing cmdline arguments with argparse, evaluate if compiler value is supported (while looking at args.travis value).
That would also help you get rid of the _compiler_choices function which is a bit confusing anyway.

Comment (jtattermusch) [inline tools/run_tests/run_tests.py]: Also, whenever you are skipping something with green color, you *must* at least print some warning.

File: tools/run_tests/run_tests.py [hunk 16]
```diff
@@ -78,6 +77,14 @@ def platform_string():
 _DEFAULT_TIMEOUT_SECONDS = 5 * 60
 
 
+# TODO (https://github./grpc/grpc/issues/7934): An object which when passed
+# as a compiler option to the script along with the travis flag, ensures that
+# the tests pass on any language. This approach will no longer be necessary
+# when the testing infrastructure would be able to support different testing
+# configurations for different branches of the repository.
+_FREE_PASS = object()
```
Comment (jtattermusch) [inline tools/run_tests/run_tests.py]: the name "FREE_PASS" could be more descriptive. 
To illustrate this, look at the `if self.args.compiler is _FREE_PASS and self.args.travis:` line pretending you don't know the context and you will be really wondering what the heck is "FREE_PASS".

How about
_UNSUPPORTED_COMPILER_SKIP_TESTS?

File: tools/run_tests/run_tests.py (hunk 11 above)
Comment (jtattermusch) [inline tools/run_tests/run_tests.py]: I still think it's very dangerous to just accept any compiler name and pretend the tests have passed. Imagine someone makes a typo while setting up the jenkins testing matrix (I can easily see someone typing "python36" instead of "python3.6" and thinking how great it is the tests are all green).

File: tools/run_tests/run_tests.py [hunk 17]
```diff
@@ -217,6 +226,8 @@ def test_specs(self):
     return sorted(out)
 
   def make_targets(self):
+    if self.args.compiler is _FREE_PASS and self.args.travis:
```
Comment (jtattermusch) [inline tools/run_tests/run_tests.py]: You've already tested for --travis in _compiler_choices, why are you testing for self.args.travis again here (and 10 other times elsewhere)?

**Comment by jtattermusch** [comment]
As you might have noticed, we changed the way we run our tests on Jenkins after #8141 has been merged (so basically the "right approach" work I pointed out in one of my previous comments is done now). Therefore, this PR has now become obsolete.

To enable/disable specific python version in a given branch, you just need to update the script
https://github.com/grpc/grpc/blob/602edbba8cf03279b3dc2e027ea07660332f806e/tools/run_tests/run_tests_matrix.py#L183 in that branch.

I'm closing this because we have a much simpler way to address the concern raised in #7934.
